fili-core/src/main/java/com/yahoo/bard/webservice/table/MetricUnionCompositeTable.java

```diff
@@ -0,0 +1,124 @@
+ * +---------+---------+---------+---------+---------+---------+
+ * |         |         |         |         |         |         |
+ * |         |         |         |         |         |         |
+ * +---------+---------+---------+---------+---------+---------+
+ * }
+ * </pre>
+ * and this joined table is backed by the <tt>MetricUnionAvailability</tt>
+ *
+ */
+public class MetricUnionCompositeTable extends BasePhysicalTable {
+    private static final Logger LOG = LoggerFactory.getLogger(MetricUnionCompositeTable.class);
+
+    /**
+     * Constructor.
+     *
+     * @param name  Name that represents set of fact table names joined together
+     * @param columns  The columns for this table
+     * @param physicalTables  A set of <tt>PhysicalTable</tt>s whose same metric schema is to be joined together. The
+     * tables will be used to construct MetricUnionAvailability, as well as to compute common/coarsest time grain among
+     * them. The <tt>PhysicalTable</tt>s needs to have mutually satisfying time grains in order to calculate the
+     * common/coarsest time grain.
+     * @param logicalToPhysicalColumnNames  Mappings from logical to physical names
+     */
+    public MetricUnionCompositeTable(
+            @NotNull TableName name,
+            @NotNull Set<Column> columns,
+            @NotNull Set<PhysicalTable> physicalTables,
+            @NotNull Map<String, String> logicalToPhysicalColumnNames
+    ) {
+        super(
+                name,
+                verifyCoarsestTimeGrain(IntervalUtils.getCoarsestTimeGrain(physicalTables), physicalTables),
+                columns,
+                logicalToPhysicalColumnNames,
+                new MetricUnionAvailability(physicalTables, columns)
+        );
+    }
+
+    /**
+     * Verifies and returns the coarsest <tt>ZonedTimeGrain</tt> if that satisfies all tables.
+     *
+     * @param coarsestGrain  The coarsest <tt>ZonedTimeGrain</tt> to be verified
+     * @param physicalTables  A set of <tt>PhysicalTable</tt>s whose <tt>ZonedTimeGrain</tt>s are checked to make sure
+     * they all satisfies with the given coarsest <tt>ZonedTimeGrain</tt>
+     *
+     * @return the coarsest <tt>ZonedTimeGrain</tt> among a set of <tt>PhysicalTables</tt>s
+     * @throws IllegalArgumentException when no PhysicalTable is provided or there is not mutually satisfying grain
+     * among the table's time grain
+     */
+    private static ZonedTimeGrain verifyCoarsestTimeGrain(
+            Optional<ZonedTimeGrain> coarsestGrain,
+            Set<PhysicalTable> physicalTables
+    ) throws IllegalArgumentException {
+        ZonedTimeGrain coarsestTimeGrain = coarsestGrain
+                .orElseThrow(() -> {
+                        String message = "At least 1 physical table needs to be provided";
+                        LOG.error(message);
+                        return new IllegalArgumentException(message);
+                });
```

Comment:
Move this check up onto the call to `IntervalUtils.getCoarsestTimeGrain(physicalTables)` in the param list to the `super(...)` call.